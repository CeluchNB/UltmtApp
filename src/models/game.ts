import { DisplayUser } from '../types/user'
import { Realm } from '@realm/react'
import { Tournament } from '../types/tournament'
import { CreateGame, Game, GameStatus, PointStats } from '../types/game'
import { DisplayTeam, GuestTeam } from '../types/team'

export class GameSchema {
    static schema: Realm.ObjectSchema = {
        name: 'Game',
        primaryKey: '_id',
        properties: {
            _id: 'string',
            creator: 'DisplayUser',
            teamOne: 'DisplayTeam',
            teamTwo: 'GuestTeam',
            teamTwoDefined: 'bool',
            scoreLimit: 'int',
            halfScore: 'int',
            startTime: 'date',
            softcapMins: 'int',
            hardcapMins: 'int',
            playersPerPoint: 'int',
            timeoutPerHalf: 'int',
            floaterTimeout: 'bool',
            tournament: 'Tournament?',
            teamOneScore: 'int',
            teamTwoScore: 'int',
            teamOnePlayers: 'DisplayUser[]',
            teamTwoPlayers: 'DisplayUser[]',
            resolveCode: 'string',
            statsPoints: 'PointStats[]',
            offline: 'bool',
            teamOneStatus: 'string',
            teamTwoStatus: 'string',
        },
    }

    _id: string
    creator: DisplayUser
    teamOne: DisplayTeam
    teamTwo: GuestTeam
    teamTwoDefined: boolean
    scoreLimit: number
    halfScore: number
    startTime: Date
    softcapMins: number
    hardcapMins: number
    playersPerPoint: number
    timeoutPerHalf: number
    floaterTimeout: boolean
    tournament?: Tournament
    teamOneScore: number
    teamTwoScore: number
    teamOnePlayers: DisplayUser[]
    teamTwoPlayers: DisplayUser[]
    resolveCode: string
    statsPoints: PointStats[]
    offline: boolean
    teamOneStatus: GameStatus
    teamTwoStatus: GameStatus

    constructor(
        game: Game,
        offline: boolean = false,
        statsPoints: PointStats[] = [],
    ) {
        this._id = game._id
        this.creator = game.creator
        this.teamOne = game.teamOne
        this.teamTwo = game.teamTwo
        this.teamTwoDefined = game.teamTwoDefined
        this.scoreLimit = game.scoreLimit
        this.halfScore = game.halfScore
        this.startTime = game.startTime
        this.softcapMins = game.softcapMins
        this.hardcapMins = game.hardcapMins
        this.playersPerPoint = game.playersPerPoint
        this.timeoutPerHalf = game.timeoutPerHalf
        this.floaterTimeout = game.floaterTimeout
        this.tournament = game.tournament
        this.teamOneScore = game.teamOneScore
        this.teamTwoScore = game.teamTwoScore
        this.teamOnePlayers = game.teamOnePlayers
        this.teamTwoPlayers = game.teamTwoPlayers
        this.resolveCode = game.resolveCode || ''
        this.statsPoints = statsPoints
        this.offline = offline
        this.teamOneStatus = game.teamOneStatus
        this.teamTwoStatus = game.teamTwoStatus
    }

    static createOfflineGame(game: CreateGame, teamOnePlayers: DisplayUser[]) {
        return {
            _id: new Realm.BSON.ObjectID().toHexString(),
            creator: game.creator,
            teamOne: game.teamOne,
            teamTwo: game.teamTwo,
            teamTwoDefined: game.teamTwoDefined,
            scoreLimit: game.scoreLimit,
            halfScore: game.halfScore,
            startTime: new Date(game.startTime),
            softcapMins: game.softcapMins,
            hardcapMins: game.hardcapMins,
            playersPerPoint: game.playersPerPoint,
            timeoutPerHalf: game.timeoutPerHalf,
            floaterTimeout: game.floaterTimeout,
            tournament: game.tournament,
            teamOneScore: 0,
            teamTwoScore: 0,
            teamOneActive: true,
            teamTwoActive: false,
            teamOnePlayers: teamOnePlayers,
            teamTwoPlayers: [],
            resolveCode: '',
            statsPoints: [],
            offline: true,
        }
    }
}
export const PointStatsSchema: Realm.ObjectSchema = {
    name: 'PointStats',
    embedded: true,
    properties: {
        _id: 'string',
        pointStats: 'PlayerPointStats[]',
    },
}

export const PlayerPointStatsSchema: Realm.ObjectSchema = {
    name: 'PlayerPointStats',
    embedded: true,
    properties: {
        _id: 'string',
        firstName: 'string',
        lastName: 'string',
        username: 'string?',
        pointsPlayed: 'int',
        goals: 'int',
        assists: 'int',
        turnovers: 'int',
        blocks: 'int',
    },
}
