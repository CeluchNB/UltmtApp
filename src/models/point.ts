import { DisplayUser } from '../types/user'
import { GuestTeam } from '../types/team'
import Point from '../types/point'
import { Realm } from '@realm/react'

export class PointSchema {
    static schema: Realm.ObjectSchema = {
        name: 'Point',
        primaryKey: '_id',
        properties: {
            _id: 'string',
            pointNumber: 'int',
            teamOnePlayers: 'DisplayUser[]',
            teamTwoPlayers: 'DisplayUser[]',
            teamOneActivePlayers: 'DisplayUser[]',
            teamTwoActivePlayers: 'DisplayUser[]',
            teamOneScore: 'int',
            teamTwoScore: 'int',
            pullingTeam: 'GuestTeam',
            receivingTeam: 'GuestTeam',
            scoringTeam: 'GuestTeam',
            teamOneActions: 'string[]',
            teamTwoActions: 'string[]',
            gameId: 'string',
            teamOneStatus: 'string',
            teamTwoStatus: 'string',
        },
    }

    _id: string
    pointNumber: number
    teamOnePlayers: DisplayUser[]
    teamTwoPlayers: DisplayUser[]
    teamOneActivePlayers: DisplayUser[]
    teamTwoActivePlayers: DisplayUser[]
    teamOneScore: number
    teamTwoScore: number
    pullingTeam: GuestTeam
    receivingTeam: GuestTeam
    scoringTeam?: GuestTeam
    teamOneActions: string[]
    teamTwoActions: string[]
    gameId: string
    teamOneStatus: string
    teamTwoStatus: string

    constructor(point: Point) {
        this._id = point._id
        this.pointNumber = point.pointNumber
        this.teamOnePlayers = point.teamOnePlayers
        this.teamTwoPlayers = point.teamTwoPlayers
        this.teamOneActivePlayers = point.teamOneActivePlayers
        this.teamTwoActivePlayers = point.teamTwoActivePlayers
        this.teamOneScore = point.teamOneScore
        this.teamTwoScore = point.teamTwoScore
        this.pullingTeam = point.pullingTeam
        this.receivingTeam = point.receivingTeam
        this.scoringTeam = point.scoringTeam
        this.teamOneActions = point.teamOneActions
        this.teamTwoActions = point.teamTwoActions
        this.gameId = point.gameId
        this.teamOneStatus = point.teamOneStatus
        this.teamTwoStatus = point.teamTwoStatus
    }
}
