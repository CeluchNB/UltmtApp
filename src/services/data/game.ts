import * as Constants from '../../utils/constants'
import { AxiosResponse } from 'axios'
import EncryptedStorage from 'react-native-encrypted-storage'
import { GuestUser } from '../../types/user'
import { throwApiError } from '../../utils/service-utils'
import { withToken } from './auth'
import { CreateGame, Game } from '../../types/game'
import {
    addGuestPlayer as networkAddGuestPlayer,
    createGame as networkCreateGame,
    searchGames as networkSearchGames,
} from '../network/game'

/**
 * Method to search games with available search and query parameters.
 * @param q search query
 * @param live filter by live games (true = returns live games, false = returns completed games, undefined = both)
 * @param after date the game occurred after
 * @param before date the game occurred before
 * @param pageSize amount of results to return
 * @param offset amount of results to skip
 * @returns list of games
 */
export const searchGames = async (
    q?: string,
    live?: boolean,
    after?: string,
    before?: string,
    pageSize?: number,
    offset?: number,
): Promise<Game[]> => {
    try {
        const response = await networkSearchGames(
            q,
            live,
            after,
            before,
            pageSize,
            offset,
        )

        const { games } = response.data

        return games
    } catch (e) {
        return throwApiError(e, Constants.SEARCH_ERROR)
    }
}

/**
 * Method to create a game
 * @param data all necessary data to create a game
 * @returns the created game
 */
export const createGame = async (data: CreateGame): Promise<Game> => {
    try {
        const response = await withToken(networkCreateGame, data)

        const { game, token } = response.data
        await EncryptedStorage.setItem('game_token', token)
        return game
    } catch (e) {
        return throwApiError(e, Constants.CREATE_GAME_ERROR)
    }
}

/**
 * Method to add a guest player to a specific team for one game
 * @param player guest player to add to a team for a game
 * @returns updated game
 */
export const addGuestPlayer = async (player: GuestUser): Promise<Game> => {
    try {
        const response = await withGameToken(networkAddGuestPlayer, player)
        const { game } = response.data
        return game
    } catch (e) {
        return throwApiError(e, Constants.ADD_GUEST_ERROR)
    }
}

export const withGameToken = async (
    networkCall: (gameToken: string, ...args: any) => Promise<AxiosResponse>,
    ...args: any
) => {
    const token = (await EncryptedStorage.getItem('game_token')) || ''
    return await networkCall(token, ...args)
}
