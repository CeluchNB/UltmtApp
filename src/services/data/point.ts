import * as Constants from '../../utils/constants'
import { GuestUser } from '../../types/user'
import Point from '../../types/point'
import { throwApiError } from '../../utils/service-utils'
import { withGameToken } from './game'
import {
    createPoint as networkCreatePoint,
    setPlayers as networkSetPlayers,
} from '../network/point'

/**
 * Method to create a point
 * @param pulling boolean - calling team pulling or not
 * @param pointNumber number of point in the game
 * @returns created point
 */
export const createPoint = async (
    pulling: boolean,
    pointNumber: number,
): Promise<Point> => {
    try {
        const response = await withGameToken(
            networkCreatePoint,
            pulling,
            pointNumber,
        )

        const { point } = response.data
        return point
    } catch (e) {
        return throwApiError(e, Constants.CREATE_POINT_ERROR)
    }
}

/**
 * Method to select the players of a specific point
 * @param pointId id of point
 * @param players array of players (length must equal game's players per point value)
 * @returns updated point
 */
export const setPlayers = async (
    pointId: string,
    players: GuestUser[],
): Promise<Point> => {
    try {
        const response = await withGameToken(
            networkSetPlayers,
            pointId,
            players,
        )
        const { point } = response.data
        return point
    } catch (e) {
        return throwApiError(e, Constants.SET_PLAYERS_ERROR)
    }
}
