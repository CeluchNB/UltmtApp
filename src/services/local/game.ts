import * as Constants from '../../utils/constants'
import AsyncStorage from '@react-native-async-storage/async-storage'
import { DisplayUser } from '../../types/user'
import { getRealm } from '../../models/realm'
import { throwApiError } from '../../utils/service-utils'
import { ActionSchema, GameSchema, PointSchema } from '../../models'
import { CreateGame, Game } from '../../types/game'

const parseGame = (schema: GameSchema): Game & { offline: boolean } => {
    return JSON.parse(
        JSON.stringify({
            _id: schema._id,
            creator: schema.creator,
            teamOne: schema.teamOne,
            teamTwo: schema.teamTwo,
            teamTwoDefined: schema.teamTwoDefined,
            scoreLimit: schema.scoreLimit,
            halfScore: schema.halfScore,
            startTime: schema.startTime,
            softcapMins: schema.softcapMins,
            hardcapMins: schema.hardcapMins,
            playersPerPoint: schema.playersPerPoint,
            timeoutPerHalf: schema.timeoutPerHalf,
            floaterTimeout: schema.floaterTimeout,
            tournament: schema.tournament,
            teamOneScore: schema.teamOneScore,
            teamTwoScore: schema.teamTwoScore,
            teamOneActive: schema.teamOneActive,
            teamTwoActive: schema.teamTwoActive,
            teamOnePlayers: schema.teamOnePlayers,
            teamTwoPlayers: schema.teamTwoPlayers,
            resolveCode: schema.resolveCode,
            points: schema.points,
            offline: schema.offline,
        }),
    )
}

/**
 * Method to determine if currently active game is offline
 * @returns boolean indicating current game is offline or not
 */
export const activeGameOffline = async (): Promise<boolean> => {
    return (await AsyncStorage.getItem('active_game_offline')) === 'true'
}

/**
 * Method to get the ObjectId of the current active game
 * @returns current active game id
 */
export const activeGameId = async (): Promise<string> => {
    return (await AsyncStorage.getItem('active_game_id')) || ''
}

/**
 * Method to set active game offline status
 * @param offline boolean
 */
export const setActiveGameOffline = async (offline: boolean) => {
    await AsyncStorage.setItem('active_game_offline', offline.toString())
}

/**
 * Method to set active game id
 * @param id string
 */
export const setActiveGameId = async (id: string) => {
    await AsyncStorage.setItem('active_game_id', id)
}

export const createOfflineGame = async (
    data: CreateGame,
    teamOnePlayers: DisplayUser[],
): Promise<string> => {
    const realm = await getRealm()

    let id: string = ''
    realm.write(() => {
        const game = realm.create<GameSchema>(
            'Game',
            GameSchema.createOfflineGame(data, teamOnePlayers),
        )
        id = game._id
    })

    return id
}

export const saveGame = async (game: Game) => {
    const realm = await getRealm()
    const gameRecord = await realm.objectForPrimaryKey<GameSchema>(
        'Game',
        game._id,
    )
    let offline = false
    if (gameRecord) {
        offline = gameRecord.offline
    }
    const schema = new GameSchema(game, offline)
    realm.write(() => {
        realm.create('Game', schema, Realm.UpdateMode.Modified)
    })
}

export const getGameById = async (
    gameId: string,
): Promise<Game & { offline: boolean }> => {
    const realm = await getRealm()
    const game = await realm.objectForPrimaryKey<GameSchema>('Game', gameId)
    if (!game) {
        return throwApiError({}, Constants.GET_GAME_ERROR)
    }
    return parseGame(game)
}

export const activeGames = async (
    userId: string,
): Promise<(Game & { offline: boolean })[]> => {
    const realm = await getRealm()
    const games = await realm.objects<GameSchema>('Game')

    return games
        .filter(g => g.creator._id === userId)
        .map(game => parseGame(game))
}

export const deleteFullGame = async (gameId: string): Promise<void> => {
    const realm = await getRealm()
    const game = await realm.objectForPrimaryKey<GameSchema>('Game', gameId)

    const points = game?.points.map(id => {
        return realm.objectForPrimaryKey<PointSchema>('Point', id)
    })

    const actions: (ActionSchema | null)[] = []
    points?.forEach(p => {
        p?.teamOneActions.forEach(id => {
            const objectId = new Realm.BSON.ObjectID(id)
            actions.push(
                realm.objectForPrimaryKey<ActionSchema>('Action', objectId),
            )
        })
        p?.teamTwoActions.forEach(id => {
            const objectId = new Realm.BSON.ObjectID(id)
            actions.push(
                realm.objectForPrimaryKey<ActionSchema>('Action', objectId),
            )
        })
    })

    realm.write(() => {
        realm.delete(actions.filter(a => a !== null))
        realm.delete(points?.filter(p => p !== null))
        realm.delete(game)
    })
}
