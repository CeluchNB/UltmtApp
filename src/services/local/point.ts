import * as Constants from '../../utils/constants'
import Point from '../../types/point'
import { Realm } from '@realm/react'
import { getLocalGameId } from './game'
import { getRealm } from '../../models/realm'
import { throwApiError } from '../../utils/service-utils'
import { GameSchema, PointSchema } from '../../models'

const parsePoint = (schema: PointSchema): Point => {
    return JSON.parse(
        JSON.stringify({
            _id: schema._id,
            pointNumber: schema.pointNumber,
            teamOnePlayers: schema.teamOnePlayers,
            teamTwoPlayers: schema.teamTwoPlayers,
            teamOneScore: schema.teamTwoScore,
            teamTwoScore: schema.teamTwoScore,
            pullingTeam: schema.pullingTeam,
            receivingTeam: schema.receivingTeam,
            scoringTeam: schema.scoringTeam,
            teamOneActive: schema.teamOneActive,
            teamTwoActive: schema.teamTwoActive,
            teamOneActions: schema.teamOneActions,
            teamTwoActions: schema.teamTwoActions,
        }),
    )
}

export const savePoint = async (point: Point) => {
    const realm = await getRealm()
    const gameId = await getLocalGameId()
    const game = await realm.objectForPrimaryKey<GameSchema>('Game', gameId)

    if (!game) {
        return throwApiError({}, Constants.GET_GAME_ERROR)
    }

    realm.write(() => {
        const rPoint = realm.create('Point', point, Realm.UpdateMode.Modified)
        game.points = [...new Set([...game.points, rPoint._id])]
    })
}

export const getPointById = async (pointId: string): Promise<Point> => {
    const realm = await getRealm()
    const point = await realm.objectForPrimaryKey<PointSchema>('Point', pointId)
    if (!point) {
        return throwApiError({}, Constants.GET_POINT_ERROR)
    }
    return parsePoint(point)
}
