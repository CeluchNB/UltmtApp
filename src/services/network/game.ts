import { CreateGame } from '../../types/game'
import { GuestUser } from '../../types/user'
import { addQueryParam } from '../../utils/service-utils'
import { API_KEY, API_URL_V1 } from 'react-native-dotenv'
import axios, { AxiosResponse } from 'axios'

export const searchGames = async (
    q?: string,
    live?: boolean,
    after?: string,
    before?: string,
    pageSize?: number,
    offset?: number,
): Promise<AxiosResponse> => {
    let url = `${API_URL_V1}/game/search`
    url = addQueryParam(url, 'q', q)
    url = addQueryParam(url, 'live', live)
    url = addQueryParam(url, 'after', after)
    url = addQueryParam(url, 'before', before)
    url = addQueryParam(url, 'pageSize', pageSize)
    url = addQueryParam(url, 'offset', offset)
    return await axios.get(url, {
        headers: { 'X-API-Key': API_KEY },
    })
}

export const createGame = async (
    token: string,
    data: CreateGame,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/game`,
        { createGameData: data },
        { headers: { 'X-API-Key': API_KEY, Authorization: `Bearer ${token}` } },
    )
}

export const addGuestPlayer = async (
    token: string,
    player: GuestUser,
): Promise<AxiosResponse> => {
    return await axios.put(
        `${API_URL_V1}/game/player/guest`,
        { player },
        { headers: { 'X-API-Key': API_KEY, Authorization: `Bearer ${token}` } },
    )
}
