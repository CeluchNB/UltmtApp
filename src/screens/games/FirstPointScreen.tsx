import BaseScreen from '../../components/atoms/BaseScreen'
import { LiveGameProps } from '../../types/navigation'
import PrimaryButton from '../../components/atoms/PrimaryButton'
import ScreenTitle from '../../components/atoms/ScreenTitle'
import { createPoint } from '../../store/reducers/features/point/livePointReducer'
import { selectGame } from '../../store/reducers/features/game/liveGameReducer'
import { useColors } from '../../hooks'
import React, { useEffect, useState } from 'react'
import { StyleSheet, Text, View } from 'react-native'
import {
    selectCreateError,
    selectCreateStatus,
    selectPoint,
} from '../../store/reducers/features/point/livePointReducer'
import { size, weight } from '../../theme/fonts'
import { useDispatch, useSelector } from 'react-redux'

const FirstPointScreen: React.FC<LiveGameProps> = ({ navigation }) => {
    const { colors } = useColors()
    const dispatch = useDispatch()
    const game = useSelector(selectGame)
    const createStatus = useSelector(selectCreateStatus)
    const createError = useSelector(selectCreateError)
    const point = useSelector(selectPoint)
    const [pulling, setPulling] = useState<boolean>()

    const onCreate = async (isPulling: boolean) => {
        setPulling(isPulling)
        dispatch(createPoint({ pulling: isPulling, pointNumber: 1 }))
    }

    useEffect(() => {
        if (createStatus === 'success') {
            navigation.navigate('SelectPlayers')
        }
    }, [createStatus, navigation, point])

    const styles = StyleSheet.create({
        title: {
            alignSelf: 'center',
            marginTop: 10,
        },
        container: {
            alignSelf: 'center',
            textAlign: 'center',
            marginTop: 10,
            width: '60%',
        },
        description: {
            color: colors.gray,
            fontSize: size.fontMedium,
            textAlign: 'center',
        },
        passcode: {
            color: colors.textPrimary,
            fontSize: size.fontLarge,
            fontWeight: weight.full,
            textAlign: 'center',
        },
        button: {
            marginTop: 10,
        },
        errorText: {
            color: colors.error,
            fontSize: size.fontFifteen,
            marginTop: 10,
        },
    })

    return (
        <BaseScreen containerWidth="80%">
            <ScreenTitle style={styles.title} title="First Point" />
            <View style={styles.container}>
                <Text style={styles.description}>Join Passcode:</Text>
                <Text style={styles.passcode}>{game.resolveCode}</Text>
            </View>
            <View style={styles.container}>
                <Text style={styles.description}>My team is</Text>
                <PrimaryButton
                    style={styles.button}
                    text="pulling"
                    loading={createStatus === 'loading' && pulling === true}
                    onPress={async () => {
                        await onCreate(true)
                    }}
                />
                <PrimaryButton
                    style={styles.button}
                    text="receiving"
                    disabled={createStatus === 'loading'}
                    loading={createStatus === 'loading' && pulling === false}
                    onPress={async () => {
                        await onCreate(false)
                    }}
                />
                {createError !== undefined && (
                    <Text style={styles.errorText}>{createError}</Text>
                )}
            </View>
        </BaseScreen>
    )
}

export default FirstPointScreen
