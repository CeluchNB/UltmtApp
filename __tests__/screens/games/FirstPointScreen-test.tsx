import * as PointData from '../../../src/services/data/point'
import { LiveGameProps } from '../../../src/types/navigation'
import FirstPointScreen from '../../../src/screens/games/FirstPointScreen'
import { NavigationContainer } from '@react-navigation/native'
import Point from '../../../src/types/point'
import { Provider } from 'react-redux'
import React from 'react'
// import { createPoint } from '../../../src/store/reducers/features/point/livePointReducer'
import { game } from '../../../fixtures/data'
import { setGame } from '../../../src/store/reducers/features/game/liveGameReducer'
import store from '../../../src/store/store'
import { fireEvent, render, waitFor } from '@testing-library/react-native'

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

const navigate = jest.fn()
const props: LiveGameProps = {
    navigation: {
        navigate,
    } as any,
    route: {} as any,
}

const spy = jest.spyOn(PointData, 'createPoint')

beforeAll(() => {
    store.dispatch(
        setGame({ ...game, startTime: '2022', tournament: undefined }),
    )
})

afterEach(() => {
    spy.mockClear()
})

it('should match snapshot with data', () => {
    const snapshot = render(
        <Provider store={store}>
            <NavigationContainer>
                <FirstPointScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )
    expect(snapshot.getByText('111111')).not.toBeNull()
    expect(snapshot.toJSON()).toMatchSnapshot()
})

it('should handle pulling press', async () => {
    spy.mockReturnValueOnce(Promise.resolve({} as Point))

    const { getByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <FirstPointScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const pullBtn = getByText('pulling')
    fireEvent.press(pullBtn)
    await waitFor(() => {
        expect(navigate).toHaveBeenCalled()
    })

    expect(spy).toHaveBeenCalledWith(true, 1)
})

it('should handle receiving press', async () => {
    spy.mockReturnValueOnce(Promise.resolve({} as Point))

    const { getByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <FirstPointScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const pullBtn = getByText('receiving')
    fireEvent.press(pullBtn)
    await waitFor(() => {
        expect(navigate).toHaveBeenCalled()
    })

    expect(spy).toHaveBeenCalledWith(false, 1)
})

it('with error', async () => {
    spy.mockReturnValueOnce(Promise.reject({ message: 'Bad stuff' }))

    const { getByText, findByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <FirstPointScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const pullBtn = getByText('pulling')
    fireEvent.press(pullBtn)
    const errorMessage = await findByText('Bad stuff')
    expect(errorMessage).not.toBeNull()
})
