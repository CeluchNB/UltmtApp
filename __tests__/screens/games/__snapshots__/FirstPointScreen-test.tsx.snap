// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot with data 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#303030",
      "height": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "width": "80%",
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#ffd600",
            "fontSize": 61.64383561643835,
            "fontWeight": "bold",
          },
          Object {
            "alignSelf": "center",
            "marginTop": 10,
          },
        ]
      }
    >
      First Point
    </Text>
    <View
      style={
        Object {
          "alignSelf": "center",
          "marginTop": 10,
          "textAlign": "center",
          "width": "60%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ababab",
            "fontSize": 41.0958904109589,
            "textAlign": "center",
          }
        }
      >
        Join Passcode:
      </Text>
      <Text
        style={
          Object {
            "color": "#ffd600",
            "fontSize": 61.64383561643835,
            "fontWeight": "900",
            "textAlign": "center",
          }
        }
      >
        111111
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "center",
          "marginTop": 10,
          "textAlign": "center",
          "width": "60%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ababab",
            "fontSize": 41.0958904109589,
            "textAlign": "center",
          }
        }
      >
        My team is
      </Text>
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "#ffd600",
            "borderColor": "transparent",
            "borderRadius": 4,
            "borderStyle": "solid",
            "borderWidth": 0,
            "elevation": 2,
            "marginTop": 10,
            "minWidth": "auto",
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0.75,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 1.5,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 4,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    Object {
                      "marginHorizontal": 8,
                    },
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    undefined,
                  ],
                ]
              }
            >
              pulling
            </Text>
          </View>
        </View>
      </View>
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "#ffd600",
            "borderColor": "transparent",
            "borderRadius": 4,
            "borderStyle": "solid",
            "borderWidth": 0,
            "elevation": 2,
            "marginTop": 10,
            "minWidth": "auto",
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0.75,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 1.5,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 4,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    Object {
                      "marginHorizontal": 8,
                    },
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    undefined,
                  ],
                ]
              }
            >
              receiving
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
